File: 백준/Gold/1753. 최단경로/최단경로.py
import heapq
import sys
input=sys.stdin.readline
INF = int(1e9)
v, e = map(int, input().split()) #v:정점, e:간선
k = int(input()) #시작 정점
graph=[[]*(v+1) for _ in range(v+1)]
for i in range(e):
    x,y,w = map(int,input().split())
    graph[x].append((y,w)) #u->v 간선 가중치 w.

distance = [INF]*(v+1) #초기값

def dijkstra(start):
    q=[]
    heapq.heappush(q,(0,start))
    distance[start] = 0

    while q:
        dist, now = heapq.heappop(q)
        if distance[now] < dist:
            continue
        for i in graph[now]:
            cost = dist + i[1]
            if cost<distance[i[0]]:
                distance[i[0]] = cost
                heapq.heappush(q,(cost,i[0]))
dijkstra(k)

for i in range(1,v+1):
    if distance[i]==INF:
        print("INF")
    else:
        print(distance[i])

File: 백준/Gold/1753. 최단경로/최단경로.java
import java.util.*;

public class Main {
    static final int INF = Integer.MAX_VALUE;

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int v = scanner.nextInt(); // 정점의 수
        int e = scanner.nextInt(); // 간선의 수
        int k = scanner.nextInt(); // 시작 정점

        List<List<Node>> graph = new ArrayList<>();
        for (int i = 0; i <= v; i++) {
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i < e; i++) {
            int x = scanner.nextInt();
            int y = scanner.nextInt();
            int w = scanner.nextInt();
            graph.get(x).add(new Node(y, w)); // u->v 간선 가중치 w.
        }

        int[] distance = new int[v + 1];
        Arrays.fill(distance, INF); // 초기값

        dijkstra(graph, distance, k);

        for (int i = 1; i <= v; i++) {
            if (distance[i] == INF) {
                System.out.println("INF");
            } else {
                System.out.println(distance[i]);
            }
        }
    }

    static class Node {
        int y;
        int weight;

        Node(int y, int weight) {
            this.y = y;
            this.weight = weight;
        }
    }

    static void dijkstra(List<List<Node>> graph, int[] distance, int start) {
        PriorityQueue<Node> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a.weight));
        pq.offer(new Node(start, 0));
        distance[start] = 0;

        while (!pq.isEmpty()) {
            Node node = pq.poll();
            int now = node.y;
            int dist = node.weight;

            if (distance[now] < dist) continue;

            for (Node next : graph.get(now)) {
                int cost = dist + next.weight;

                if (cost < distance[next.y]) {
                    distance[next.y] = cost;
                    pq.offer(new Node(next.y, cost));
                }
            }
        }
    }
}
